### PR TITLE
#182210048 logout functionality

### PR DESCRIPTION
**Pull request by munezerobagira**
#### What does this PR do?

Users should be able to log out of the application

#### Description of Task to be completed?
- [x]  delete the user session (logout the user)
- [x] store the login information (ip, device type and token)
#### How should this be manually tested?
1. Clone the repository
2. run the server `npm run dev`
3. open swagger documentations "localhost:your-prefered-port/api-docs"
4. Test the following endpoints
- "/api/auth/signout"
- GET "/api/auth/session" get request
- DELETE "/api/auth/session/<sessionsid>"

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[ Finishes #182210048]
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

### CODE REVIEW

File: src/controllers/Authcontrollers.js [hunk 1]
```diff
@@ -40,24 +42,32 @@ class Auth {
...
     try {
       const { password, email } = req.body;
 
       const user = await checkUser(email);
       if (!user) {
-        return errorResponse(res, 404, `User not found!`);
+        return errorResponse(res, 404, "User not found!");
+      }
+      if (!checkPassword(password, user.password)) {
+        return errorResponse(res, 409, "Invalid credentials");
```
Comment (habaumugisha1) [inline src/controllers/Authcontrollers.js]: it shouldn't be 409 it can be 403 or 401

File: src/controllers/Authcontrollers.js [hunk 2]
```diff
@@ -67,39 +77,86 @@ class Auth {
       );
     }
   }
-  
+  static async signout(req, res) {
+    try {
+      if (!req.user || !req.headers["x-auth-token"])
```
Comment (habaumugisha1) [inline src/controllers/Authcontrollers.js]: I think we are using 
```req.headers["authorization"]```

Comment (munezerobagira) [reply src/controllers/Authcontrollers.js]: Honore used x-auth-token

Comment (munezerobagira) [reply src/controllers/Authcontrollers.js]: do I  have to change it to Authorization?

File: src/middlewares/isAuthenticated.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import successResponse from "../utils/success";
+import { User } from "../database/models";
+import Protection from "./hash";
+// eslint-disable-next-line consistent-return
+
+const isAuthenticated = async (req, res, next) => {
+  try {
+    const token = req.headers["x-auth-token"];
+    if (!token)
+      return successResponse(res, 409, "Access denied. No token provided!");
```
Comment (habaumugisha1) [inline src/middlewares/isAuthenticated.js]: I think this is errorResponse

File: src/controllers/Authcontrollers.js (hunk 1 above)
Comment (munezerobagira) [reply src/controllers/Authcontrollers.js]: working on it

File: src/middlewares/isAuthenticated.js (hunk 3 above)
Comment (munezerobagira) [reply src/middlewares/isAuthenticated.js]: working on it

**Review by pacomesimon — COMMENTED** [review]
I run this feature on my PC and it's working good. On my PC, the written tests are also passing, but I am failing to figure out why there are few tests that are not passing on circle CI.

File: src/controllers/Authcontrollers.js (hunk 2 above)
Comment (habaumugisha1) [reply src/controllers/Authcontrollers.js]: Okay, leave it like that

**Review by pacomesimon — APPROVED** [review]

**Pull request merged**